extmod/os_dupterm.c

```diff
@@ -148,29 +148,41 @@ int mp_os_dupterm_rx_chr(void) {
     return -1;
 }
 
-void mp_os_dupterm_tx_strn(const char *str, size_t len) {
+int mp_os_dupterm_tx_strn(const char *str, size_t len) {
+    // Returns the minimum successful write length, or -1 if no write is attempted.
```

Comment:
Maybe for simplicity it can return `len` if no write is attempted?  Because technically it's true, that the whole write did succeed.  That should simplify the logic for most callers of this function.

And then the return type can be `size_t` and match the type of `len`.

Reply:
This was done to keep logic in line with the logic in `ports/nrf/drivers/bluetooth/ble_uart.c` and `if (MP_STATE_PORT(board_stdio_uart) == NULL)` in `nrf/mphalport.c`, where @projectgus and me agreed that "no connection == no bytes written" was logical, see https://github.com/micropython/micropython/pull/11850#discussion_r1370873223 above. If the dupterm outputs are all unconnected, it would not be logical to report that bytes were written.

(Possible separate issue: shouldn't `mp_hal_stdout_tx_strn()` in `blue_uart.c` call dupterm?)

Reply:
Is there a place where the caller performs different logic based on `0` or `-1` being returned?

It seems like in most places the caller's logic is if `if (ret < 0) ret = 0;`, in which case it seems like these functions can return 0 in both cases for "zero bytes written".

But I feel like I might be missing a nuance here?

Reply:
There is a nuance, and there are places where a return value of `0` is handled differently than `-1`. The logic is: if no write attempt had to be made in `mp_os_dupterm_tx_strn()`, it is basically a void call and it indicates nothing was done and hence nothing failed/succeeded, and the "bytes written" result of `mp_os_dupterm_tx_strn()` should not even be considered. However, if an attempt was made, it is of interest whether it wrote any bytes at all.

See e.g. `mp_hal_stdout_tx_strn()` in `ports/stm32/mphalport.c`:

```
MP_WEAK mp_uint_t mp_hal_stdout_tx_strn(const char *str, size_t len) {
    mp_uint_t ret = len;
    bool did_write = false;
    if (MP_STATE_PORT(pyb_stdio_uart) != NULL) {
        uart_tx_strn(MP_STATE_PORT(pyb_stdio_uart), str, len);
        did_write = true;
    }
    #if 0 && defined(USE_HOST_MODE) && MICROPY_HW_HAS_LCD
    lcd_print_strn(str, len);
    #endif
    int dupterm_res = mp_os_dupterm_tx_strn(str, len);
    if (dupterm_res >= 0) {
        did_write = true;
        ret = MIN((mp_uint_t)dupterm_res, ret);
    }
    return did_write?ret:0;
}
```
In case the `pyb_stdio_uart` write succeeds, we have `ret == len`. If then `mp_os_dupterm_tx_strn()` would have nothing to do (no outputs so nothing to attempt) and would return zero, then this function `mp_hal_stdout_tx_strn()` would return zero as well, indication something went wrong, while everything would actually be fine.

I think there is no shortcut here for `mp_os_dupterm_tx_strn()`, it basically has three main results: no write attempted, write attempted but (partly) failed, write succeeded completely. Any simplification there would require simplification in the higher level logic. One overall simplification could be to rework the entire thing and create a single list of outputs, i.e. make every output a "dupterm" output. But that would be a lot of work on multiple levels, with marginal benefits.

Reply:
That makes sense to me, thanks for the detailed explanation. I agree the approach in this PR is the best way to proceed.

Will leave this for @dpgeorge to pick up when he has time.

Reply:
> I think there is no shortcut here for `mp_os_dupterm_tx_strn()`, it basically has three main results: no write attempted, write attempted but (partly) failed, write succeeded completely.

Yes, I agree, it needs these three distinct return scenarios.